vim-engine/src/main/kotlin/com/maddyhome/idea/vim/api/VimMotionGroup.kt

```diff
@@ -15,116 +15,117 @@ import com.maddyhome.idea.vim.common.TextRange
 import com.maddyhome.idea.vim.handler.Motion
 
 interface VimMotionGroup {
-  var lastFTCmd: TillCharacterMotionType
-  var lastFTChar: Char
+  /**
+   * Move the caret to the given buffer offset.
+   *
+   * Also updates/resets selection, updates visual block, remembers last vertical position, etc.
+   */
+  fun moveCaret(editor: VimEditor, caret: VimCaret, offset: Int)
 
+  // Note that the following methods require the caret to access the intended vertical position, such as "end of line"
+  fun getOffsetOfHorizontalMotion(editor: VimEditor, caret: VimCaret, count: Int, allowPastEnd: Boolean): Int
   fun getVerticalMotionOffset(editor: VimEditor, caret: VimCaret, count: Int): Int
+
+  // TODO: Consider naming. These don't move the caret, but calculate offsets. Also consider returning Motion
+
+  // Move caret to buffer line
+  fun moveCaretToLineStart(editor: VimEditor, caret: VimCaret): Int
+  fun moveCaretToLineStart(editor: VimEditor, line: Int): Int
+  fun moveCaretToLineStartSkipLeading(editor: VimEditor, caret: VimCaret): Int
+  fun moveCaretToLineStartSkipLeading(editor: VimEditor, line: Int): Int
+  fun moveCaretToLineStartSkipLeadingOffset(editor: VimEditor, caret: VimCaret, linesOffset: Int): Int
+  fun moveCaretToLineWithStartOfLineOption(editor: VimEditor, logicalLine: Int, caret: VimCaret): Int
   fun moveCaretToLineEnd(editor: VimEditor, caret: VimCaret): Int
   fun moveCaretToLineEnd(editor: VimEditor, line: Int, allowPastEnd: Boolean): Int
-  fun moveCaretToLineStart(editor: VimEditor, line: Int): Int
+  fun moveCaretToLineEndOffset(editor: VimEditor, caret: VimCaret, cntForward: Int, allowPastEnd: Boolean): Int
+  fun moveCaretToLineEndSkipLeadingOffset(editor: VimEditor, caret: VimCaret, linesOffset: Int): Int
+  fun moveCaretToLineWithSameColumn(editor: VimEditor, logicalLine: Int, caret: VimCaret): Int
+  fun moveCaretToLinePercent(editor: VimEditor, caret: VimCaret, count: Int): Int
+
+  // Move caret to (IntelliJ visual) line relative to the bounds of the display (aka window)
+  // (This describes what these functions *currently* do, not what they are *supposed* to do)
+  // TODO: These should move to the Vim logical line at the top/bottom/middle of the display
+  fun moveCaretToFirstScreenLine(editor: VimEditor, caret: VimCaret, count: Int, normalizeToScreen: Boolean): Int
+  fun moveCaretToLastScreenLine(editor: VimEditor, caret: VimCaret, count: Int, normalizeToScreen: Boolean): Int
+  fun moveCaretToMiddleScreenLine(editor: VimEditor, caret: VimCaret): Int
+
+  // Move caret to buffer column
+  fun moveCaretToColumn(editor: VimEditor, caret: VimCaret, count: Int, allowEnd: Boolean): Motion
+
+  // Move caret to column relative to the bounds of the display (aka window)
+  fun moveCaretToCurrentDisplayLineStart(editor: VimEditor, caret: VimCaret): Motion
```

Comment:
Just sharing plans: I also would like to get rid of the "move" inside such method because it looks like this method does some modification while it's not

Reply:
I haven't been able to come up with a good name for this though - `calculateMotionToCurrentDisplayLineStart` or `motionToCurrentDisplayLineStart`? Or even more simple, because it's always called through `motion` - `injector.motion.toCurrentDisplayLineStart`?